### PR TITLE
Allow interaction with ray client service via Route from outside of OCP cluster

### PR DESCRIPTION
**Pull request by tedhtchang**
How to test:
```
# clone the branch
git clone --single-branch --branch ray-loca-interaction https://github.com/tedhtchang/codeflare-sdk.git
cd codeflare-sdk/

# create python virtual env with [pyenv](https://github.com/pyenv/pyenv#automatic-installer)
pyenv install 3.8.13
pyenv virtualenv 3.8.13 localinter
pyenv local localinter

# Install python packages
pip install -r requirements.txt
pip install -r requirements-dev.txt
pip install jupyterlab
pip install -e .

# Run a sameple noteook and then Run -> Run all cells
jupyter lab demo-notebooks/interactive/local_interactive.ipynb

# clean up
pyenv uninstall localinter
# delete the cloned repo
```
Expect something
image

### CODE REVIEW

**Comment by tedhtchang** [comment]
@MichaelClifford I passed all the tests. looks good. I also renamed the example notebook.

**Comment by MichaelClifford** [comment]
Thanks for adding more tests @tedhtchang 

LGTM

**Review by jbusche — APPROVED** [review]
LGTM, thank you Ted!
Note, for testing, there were two items that were different from what was listed:
1.  Had to install `pip install -r requirements.txt` in addition to the requirements-dev.txt in order to get openshift installed in my pyenv
2. The notebook file is named `local_interactive.ipynb`

**Comment by jbusche** [comment]
@tedhtchang, with the latest code with the Ray Operator v0.5.0, the worker is now able to start all 3 init containers and that's working ok.  But when I try to do the `import ray` step I'm getting:
```
File ~/.pyenv/versions/3.8.13/envs/localinter/lib/python3.8/site-packages/ray/util/client/worker.py:260, in Worker._connect_channel(self, reconnecting)
    252 if log_once("ray_client_security_groups"):
    253     warnings.warn(
    254         "Ray Client connection timed out. Ensure that "
    255         "the Ray Client port on the head node is reachable "
   (...)
    258         "more information."
    259     )
--> 260 raise ConnectionError("ray client connection timeout")

ConnectionError: ray client connection timeout
```

**Comment by tedhtchang** [comment]
Tested on my own mac and another Ubuntu VM from scratch. Both env seems to work. @MichaelClifford @Maxusmusti Could you test again?

**Comment by jbusche** [comment]
I tried one last time last night and local interaction worked on my cluster to `default` namespace from my Mac.  I'm not certain why it worked this time, but perhaps a new cluster and/or some local changes on my Mac... not sure.
Was also trying to test from an Ubuntu image, but I ran out of time trying to setup my Ubuntu with the proper prereqs.  Last thing Ted wanted me to try was a different namespace other than `default`.   Will post update here once my cluster rebuilds and I can test.

**Comment by jbusche** [comment]
OK - tried on the IBM network with both `default` and `jim` namespaces - looked good!

**Review by jbusche — DISMISSED** [review]
LGTM Ted, passing the tests in two namespaces
Thank you!

**Review by jbusche — DISMISSED** [review]
LGTM
Tested in both default and non-default namespaces... looks good.

**Review by Maxusmusti — DISMISSED** [review]
LGTM

**Review by MichaelClifford — APPROVED** [review]
LGTM! 

Tested PR on cluster with KubeRay v0.5.0 and works as expected.

File: src/codeflare_sdk/utils/generate_yaml.py [hunk 1]
```diff
@@ -191,6 +202,82 @@ def update_nodes(
...
+    ][0]["env"][1]["value"] = "1"
+    # update_init_container
+    command = item["generictemplate"]["spec"]["headGroupSpec"]["template"]["spec"][
+        "initContainers"
+    ][0].get("command")[2]
+
+    command = command.replace("deployment-name", cluster_name)
+
+    server_name = (
+        oc.whoami("--show-server").split(":")[1].split("//")[1].replace("api", "apps")
```
Comment (Maxusmusti) [inline src/codeflare_sdk/utils/generate_yaml.py:244]: Will eventually have to be changed to kubernetes client, can be done at same time as `cluster.local_client_url()` though, no need in this PR (maybe add a note in the new issue #173)

File: src/codeflare_sdk/utils/generate_yaml.py [hunk 2]
```diff
@@ -191,6 +202,82 @@ def update_nodes(
...
+    ]["spec"]["containers"][0]["volumeMounts"]
+    del resources["GenericItems"][0]["generictemplate"]["spec"]["workerGroupSpecs"][0][
+        "template"
+    ]["spec"]["initContainers"][1]
+    del resources["GenericItems"][3]  # rayclient route
+    del resources["GenericItems"][2]  # ca-secret
+
+
+def delete_route():
+    print("Delete Rayclient Route")
```
Comment (Maxusmusti) [inline src/codeflare_sdk/utils/generate_yaml.py:278]: Just noticed, does this function do anything / is there a reason for its existence?

**Review by Maxusmusti — COMMENTED** [review]
Looks good! Just two quick comments before I merge

File: src/codeflare_sdk/utils/generate_yaml.py (hunk 2 above)
Comment (tedhtchang) [reply src/codeflare_sdk/utils/generate_yaml.py:278]: no it doesn't have any function probably just some pseudo code I wrote before.

File: src/codeflare_sdk/utils/generate_yaml.py (hunk 1 above)
Comment (tedhtchang) [reply src/codeflare_sdk/utils/generate_yaml.py:244]: noted.

**Review by Maxusmusti — APPROVED** [review]
LGTM

**Pull request merged**